p2p/transport.go

```diff
@@ -0,0 +1,579 @@
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/tendermint/tendermint/config"
+	crypto "github.com/tendermint/tendermint/crypto"
+	cmn "github.com/tendermint/tendermint/libs/common"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/p2p/conn"
+)
+
+const (
+	defaultDialTimeout      = time.Second
+	defaultFilterTimeout    = 5 * time.Second
+	defaultHandshakeTimeout = 3 * time.Second
+)
+
+// Transport-level errors.
+var (
+	ErrTransportFilterTimeout = errors.New("peer filter timeout")
+)
+
+// accept is the container to carry the upgraded connection and NodeInfo from an
+// asynchronously running routine to the Accept method.
+type accept struct {
+	conn     net.Conn
+	nodeInfo NodeInfo
+	err      error
+}
+
+// peerConfig is used to bundle data we need to fully setup a Peer with an
+// MConn, provided by the caller of Accept and Dial.
+type peerConfig struct {
+	chDescs              []*conn.ChannelDescriptor
+	onPeerError          func(Peer, interface{})
+	outbound, persistent bool
+	reactorsByCh         map[byte]Reactor
+}
+
+// Transport emits and connects to Peers. The implementation of Peer is left to
+// the transport. Each transport is also responsible to filter establishing
+// peers specific to its domain.
+type Transport interface {
+	// Accept returns a newly connected Peer.
+	Accept(peerConfig) (Peer, error)
```

Comment:
This is temporrary, I will make it more clear in the comment on `peerConfig`. The reason it's still present is that `*peer` depends on `ch` and `reactor` info and the switch needs to have the error callback hooked up. It will all vanish when we make reactor setup less dynamic and have `PeerBehaviour` as concept esablished.